endorser/api/services/allow_lists.py

```diff
@@ -0,0 +1,58 @@
+import logging
+from typing import TypeVar
+from api.db.models.base import BaseModel
+from sqlalchemy.exc import IntegrityError
+from sqlalchemy import select, delete
+from sqlalchemy.ext.asyncio import AsyncSession
+from psycopg2.errors import UniqueViolation
+from api.endpoints.models.endorse import (
+    EndorseTransactionState,
+    db_to_txn_object,
+)
+from api.db.models.endorse_request import EndorseRequest
+from api.services.auto_state_handlers import is_endorsable_transaction
+from api.services.endorse import endorse_transaction
+from api.db.errors import AlreadyExists
+
+
+logger = logging.getLogger(__name__)
+
+
+async def updated_allowed(db: AsyncSession) -> None:
+    try:
+        q = select(EndorseRequest).where(
+            EndorseRequest.state == EndorseTransactionState.request_received
+        )
+        result = await db.execute(q)
+        db_txns: list[EndorseRequest] = result.scalars().all()
+        for txn in db_txns:
+            transaction = db_to_txn_object(txn, acapy_txn=None)
+            logger.debug(
+                f">>> from updated_allowed: the current transaction is {transaction}"
+            )
+            was_allowed = await is_endorsable_transaction(db, transaction)
+            logger.debug(f">>> from updated_allowed: this was allowed? {was_allowed}")
+            if was_allowed:
+                logger.debug(
+                    f">>> from updated_allowed: endorsing transaction: {transaction}"
+                )
+                await endorse_transaction(db, transaction)
+    except Exception as e:
+        # try to retrieve and print text on error
+        logger.error(f"Failed to update pending transactions {e}")
+
+
+B = TypeVar("B", bound=BaseModel)
+
+
+async def add_to_allow_list(db: AsyncSession, a: B) -> B:
```

Comment:
Any reason to not use `BaseModel` directly rater than type-bind `B` to it?
Wondering about readability in the longer term, when we get back to the code and forgot what this is supposed to do 😅

Reply:
Just to make mypy/pyright happy. This ensures that the class of `a` is identical to the return type of this function and not another class deriving from BaseModel. For example if we changed the function to
```python
async def add_to_allow_list(db: AsyncSession, a: BaseModel) -> BaseModel:
```
You get this error when calling `add_to_allow_list` in allow.py
```
Pyright [reportReturnType]: Expression of type "BaseModel" is incompatible with return type "AllowedCredentialDefinition"
```
While not as intuitive it's the proper way to enforce this